File: src/lib/api/conversationImport.ts
import { AppServerClient } from "@/lib/api/appServer";
import {
  CONVERSATION_IMPORT_JOB_PHASES,
  CONVERSATION_IMPORT_JOB_STATUSES,
  CONVERSATION_IMPORT_SOURCE_CLIENTS,
  CONVERSATION_IMPORT_SOURCE_STATUSES,
  CONVERSATION_IMPORT_THREAD_STATUSES,
  METHOD_CONVERSATION_IMPORT_JOB_READ,
  METHOD_CONVERSATION_IMPORT_SOURCE_SCAN,
  METHOD_CONVERSATION_IMPORT_THREAD_COMMIT,
  METHOD_CONVERSATION_IMPORT_THREAD_PREVIEW,
  type ConversationImportJob,
  type ConversationImportJobPhase,
  type ConversationImportJobReadParams,
  type ConversationImportJobReadResponse,
  type ConversationImportJobStatus,
  type ConversationImportSourceClient,
  type ConversationImportSourceProvenance,
  type ConversationImportSourceScanParams,
  type ConversationImportSourceScanResponse,
  type ConversationImportSourceStatus,
  type ConversationImportThreadCommitParams,
  type ConversationImportThreadCommitResponse,
  type ConversationImportThreadCommitStartResponse,
  type ConversationImportThreadStatus,
  type ConversationImportThreadPreviewParams,
  type ConversationImportThreadPreviewResponse,
  type ImportedThreadSummary,
} from "../../../packages/app-server-client/src/protocol";

export type {
  ConversationImportJob,
  ConversationImportJobPhase,
  ConversationImportJobReadParams,
  ConversationImportJobReadResponse,
  ConversationImportJobStatus,
  ConversationImportSourceProvenance,
  ConversationImportSourceClient,
  ConversationImportSourceScanParams,
  ConversationImportSourceScanResponse,
  ConversationImportSourceStatus,
  ConversationImportThreadStatus,
  ConversationImportThreadCommitParams,
  ConversationImportThreadCommitResponse,
  ConversationImportThreadCommitStartResponse,
  ConversationImportThreadPreviewParams,
  ConversationImportThreadPreviewResponse,
  ImportedThreadSummary,
};

type ConversationImportAppServerClient = Pick<AppServerClient, "request">;

const SOURCE_CLIENT_VALUES = new Set<string>(
  CONVERSATION_IMPORT_SOURCE_CLIENTS,
);
const SOURCE_STATUS_VALUES = new Set<string>(
  CONVERSATION_IMPORT_SOURCE_STATUSES,
);
const THREAD_STATUS_VALUES = new Set<string>(
  CONVERSATION_IMPORT_THREAD_STATUSES,
);
const JOB_STATUS_VALUES = new Set<string>(CONVERSATION_IMPORT_JOB_STATUSES);
const JOB_PHASE_VALUES = new Set<string>(CONVERSATION_IMPORT_JOB_PHASES);

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function isOptionalString(value: unknown): value is string | undefined {
  return value === undefined || typeof value === "string";
}

function isConversationImportSourceClient(
  value: unknown,
): value is ConversationImportSourceClient {
  return typeof value === "string" && SOURCE_CLIENT_VALUES.has(value);
}

function isConversationImportSourceStatus(
  value: unknown,
): value is ConversationImportSourceStatus {
  return typeof value === "string" && SOURCE_STATUS_VALUES.has(value);
}

function isConversationImportThreadStatus(
  value: unknown,
): value is ConversationImportThreadStatus {
  return typeof value === "string" && THREAD_STATUS_VALUES.has(value);
}

function isAgentSession(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.sessionId === "string" &&
    typeof value.threadId === "string" &&
    typeof value.appId === "string" &&
    isOptionalString(value.workspaceId) &&
    typeof value.status === "string" &&
    typeof value.createdAt === "string" &&
    typeof value.updatedAt === "string"
  );
}

function isImportedThreadSummary(
  value: unknown,
): value is ImportedThreadSummary {
  if (!isRecord(value)) {
    return false;
  }
  return (
    isConversationImportSourceClient(value.sourceClient) &&
    typeof value.sourceThreadId === "string" &&
    isOptionalString(value.title) &&
    isOptionalString(value.createdAt) &&
    isOptionalString(value.updatedAt) &&
    isOptionalString(value.cwd) &&
    isOptionalString(value.source) &&
    isOptionalString(value.modelProvider) &&
    typeof value.archived === "boolean" &&
    isOptionalString(value.sourcePath) &&
    isOptionalString(value.importJobId) &&
    isConversationImportThreadStatus(value.importStatus) &&
    (value.metadata === undefined ||
      value.metadata === null ||
      typeof value.metadata !== "function")
  );
}

function isConversationImportJobStatus(
  value: unknown,
): value is ConversationImportJobStatus {
  return typeof value === "string" && JOB_STATUS_VALUES.has(value);
}

function isConversationImportJobPhase(
  value: unknown,
): value is ConversationImportJobPhase {
  return typeof value === "string" && JOB_PHASE_VALUES.has(value);
}

function isOptionalNumber(value: unknown): value is number | undefined {
  return value === undefined || typeof value === "number";
}

function isSourceProvenance(
  value: unknown,
): value is ConversationImportSourceProvenance {
  if (value === undefined) {
    return true;
  }
  return (
    isRecord(value) &&
    isConversationImportSourceClient(value.sourceClient) &&
    isOptionalString(value.sourceThreadId) &&
    isOptionalString(value.sourcePath) &&
    isOptionalString(value.sourceEventType) &&
    isOptionalNumber(value.sourceEventSeq) &&
    isOptionalString(value.sourcePayloadType) &&
    isOptionalString(value.sourceCallId) &&
    isOptionalString(value.sourceRole) &&
    isOptionalString(value.sourceChannel)
  );
}

function isPreviewMessage(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.role === "string" &&
    typeof value.text === "string" &&
    Array.isArray(value.attachments) &&
    value.attachments.every(
      (attachment) =>
        isRecord(attachment) &&
        typeof attachment.kind === "string" &&
        isOptionalString(attachment.uri) &&
        (attachment.metadata === undefined ||
          attachment.metadata === null ||
          typeof attachment.metadata !== "function"),
    ) &&
    typeof value.truncated === "boolean" &&
    typeof value.omittedBytes === "number" &&
    isOptionalString(value.timestamp) &&
    isOptionalString(value.sourceType) &&
    isSourceProvenance(value.provenance)
  );
}

function isPreviewEvent(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.kind === "string" &&
    isOptionalString(value.timestamp) &&
    isOptionalString(value.label) &&
    isSourceProvenance(value.provenance)
  );
}

function isPreviewDryRun(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.willCreateSession === "boolean" &&
    typeof value.willAppendToExistingSession === "boolean" &&
    typeof value.willImportMessages === "number" &&
    typeof value.willImportTurns === "number" &&
    typeof value.willImportTimelineItems === "number" &&
    typeof value.willImportAttachments === "number" &&
    typeof value.unsupportedItems === "number"
  );
}

function isFidelitySummary(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.messages === "number" &&
    typeof value.reasoning === "number" &&
    typeof value.tools === "number" &&
    typeof value.commands === "number" &&
    typeof value.patches === "number" &&
    typeof value.approvals === "number" &&
    typeof value.mcp === "number" &&
    typeof value.webSearch === "number" &&
    typeof value.attachments === "number" &&
    typeof value.unsupported === "number" &&
    typeof value.provenanceOnly === "number" &&
    typeof value.budgetDropped === "number"
  );
}

function isPreviewSummary(value: unknown): boolean {
  return (
    isRecord(value) &&
    typeof value.lineCount === "number" &&
    typeof value.messageCount === "number" &&
    typeof value.rolloutEventItems === "number" &&
    typeof value.unsupportedCount === "number" &&
    isPreviewDryRun(value.dryRun) &&
    isFidelitySummary(value.fidelity) &&
    typeof value.truncated === "boolean" &&
    Array.isArray(value.warnings) &&
    value.warnings.every((warning) => typeof warning === "string")
  );
}

function assertConversationImportSourceScanResponse(
  value: unknown,
): asserts value is ConversationImportSourceScanResponse {
  if (!isRecord(value) || !isRecord(value.source)) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_SOURCE_SCAN} did not return a source scan response`,
    );
  }
  const { source } = value;
  if (
    !isConversationImportSourceClient(source.sourceClient) ||
    !isConversationImportSourceStatus(source.status) ||
    !isOptionalString(source.sourceRoot) ||
    typeof source.readable !== "boolean" ||
    typeof source.threadCount !== "number" ||
    typeof source.sourceHomeExists !== "boolean" ||
    typeof source.stateDbReadable !== "boolean" ||
    typeof source.rolloutFileCount !== "number" ||
    !isOptionalString(source.indexedAt) ||
    !isOptionalString(source.statePath) ||
    !isOptionalString(source.message) ||
    !Array.isArray(value.threads) ||
    !value.threads.every(isImportedThreadSummary) ||
    !isOptionalString(value.nextCursor)
  ) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_SOURCE_SCAN} returned an invalid source scan shape`,
    );
  }
}

function assertConversationImportThreadPreviewResponse(
  value: unknown,
): asserts value is ConversationImportThreadPreviewResponse {
  if (
    !isRecord(value) ||
    !isRecord(value.source) ||
    !isImportedThreadSummary(value.thread) ||
    !isPreviewSummary(value.summary) ||
    !Array.isArray(value.messages) ||
    !value.messages.every(isPreviewMessage) ||
    !Array.isArray(value.events) ||
    !value.events.every(isPreviewEvent)
  ) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_THREAD_PREVIEW} returned an invalid thread preview shape`,
    );
  }
  assertConversationImportSourceScanResponse({
    source: value.source,
    threads: [],
  });
}

function assertConversationImportThreadCommitResponse(
  value: unknown,
): asserts value is ConversationImportThreadCommitResponse {
  if (
    !isRecord(value) ||
    !isAgentSession(value.session) ||
    !isImportedThreadSummary(value.thread) ||
    !isPreviewSummary(value.summary) ||
    typeof value.importedMessages !== "number" ||
    typeof value.importedTurns !== "number" ||
    typeof value.canContinue !== "boolean" ||
    !Array.isArray(value.warnings) ||
    !value.warnings.every((warning) => typeof warning === "string")
  ) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_THREAD_COMMIT} returned an invalid thread commit shape`,
    );
  }
}

function isConversationImportJob(
  value: unknown,
): value is ConversationImportJob {
  if (
    !isRecord(value) ||
    typeof value.jobId !== "string" ||
    !isConversationImportSourceClient(value.sourceClient) ||
    !isOptionalString(value.sourceThreadId) ||
    !isConversationImportJobStatus(value.status) ||
    !isRecord(value.progress) ||
    !isConversationImportJobPhase(value.progress.phase) ||
    typeof value.progress.completedItems !== "number" ||
    typeof value.progress.totalItems !== "number" ||
    typeof value.progress.completedTurns !== "number" ||
    typeof value.progress.totalTurns !== "number" ||
    !isOptionalString(value.error) ||
    typeof value.createdAt !== "string" ||
    typeof value.updatedAt !== "string"
  ) {
    return false;
  }
  if (value.result !== undefined) {
    try {
      assertConversationImportThreadCommitResponse(value.result);
    } catch {
      return false;
    }
  }
  return true;
}

function assertConversationImportThreadCommitStartResponse(
  value: unknown,
): asserts value is ConversationImportThreadCommitStartResponse {
  if (!isRecord(value) || !isConversationImportJob(value.job)) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_THREAD_COMMIT} did not return an import job`,
    );
  }
}

function assertConversationImportJobReadResponse(
  value: unknown,
): asserts value is ConversationImportJobReadResponse {
  if (!isRecord(value) || !isConversationImportJob(value.job)) {
    throw new Error(
      `${METHOD_CONVERSATION_IMPORT_JOB_READ} did not return an import job`,
    );
  }
}

export async function scanConversationImportSource(
  params: ConversationImportSourceScanParams = {},
  appServerClient: ConversationImportAppServerClient = new AppServerClient(),
): Promise<ConversationImportSourceScanResponse> {
  const response =
    await appServerClient.request<ConversationImportSourceScanResponse>(
      METHOD_CONVERSATION_IMPORT_SOURCE_SCAN,
      params,
    );
  assertConversationImportSourceScanResponse(response.result);
  return response.result;
}

export async function previewConversationImportThread(
  params: ConversationImportThreadPreviewParams,
  appServerClient: ConversationImportAppServerClient = new AppServerClient(),
): Promise<ConversationImportThreadPreviewResponse> {
  const response =
    await appServerClient.request<ConversationImportThreadPreviewResponse>(
      METHOD_CONVERSATION_IMPORT_THREAD_PREVIEW,
      params,
    );
  assertConversationImportThreadPreviewResponse(response.result);
  return response.result;
}

export async function commitConversationImportThread(
  params: ConversationImportThreadCommitParams,
  appServerClient: ConversationImportAppServerClient = new AppServerClient(),
): Promise<ConversationImportThreadCommitStartResponse> {
  const response =
    await appServerClient.request<ConversationImportThreadCommitStartResponse>(
      METHOD_CONVERSATION_IMPORT_THREAD_COMMIT,
      params,
    );
  assertConversationImportThreadCommitStartResponse(response.result);
  return response.result;
}

export async function readConversationImportJob(
  params: ConversationImportJobReadParams,
  appServerClient: ConversationImportAppServerClient = new AppServerClient(),
): Promise<ConversationImportJobReadResponse> {
  const response =
    await appServerClient.request<ConversationImportJobReadResponse>(
      METHOD_CONVERSATION_IMPORT_JOB_READ,
      params,
    );
  assertConversationImportJobReadResponse(response.result);
  return response.result;
}

export interface WaitForConversationImportJobOptions {
  intervalMs?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
  onProgress?: (job: ConversationImportJob) => void;
}

export async function waitForConversationImportJob(
  initialJob: ConversationImportJob,
  options: WaitForConversationImportJobOptions = {},
  appServerClient: ConversationImportAppServerClient = new AppServerClient(),
): Promise<ConversationImportThreadCommitResponse> {
  const intervalMs = Math.max(100, options.intervalMs ?? 250);
  const timeoutMs = Math.max(intervalMs, options.timeoutMs ?? 30 * 60_000);
  const deadline = Date.now() + timeoutMs;
  let job = initialJob;

  while (true) {
    options.onProgress?.(job);
    if (job.status === "completed") {
      if (!job.result) {
        throw new Error("Conversation import completed without a result");
      }
      return job.result;
    }
    if (job.status === "failed") {
      throw new Error(job.error || "Conversation import failed");
    }
    if (options.signal?.aborted) {
      throw new DOMException(
        "Conversation import wait was aborted",
        "AbortError",
      );
    }
    if (Date.now() >= deadline) {
      throw new Error(`Conversation import job timed out: ${job.jobId}`);
    }

    await waitForPoll(intervalMs, options.signal);
    job = (
      await readConversationImportJob({ jobId: job.jobId }, appServerClient)
    ).job;
  }
}

function waitForPoll(intervalMs: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const handleAbort = () => {
      window.clearTimeout(timeout);
      reject(
        new DOMException("Conversation import wait was aborted", "AbortError"),
      );
    };
    const timeout = window.setTimeout(() => {
      signal?.removeEventListener("abort", handleAbort);
      resolve();
    }, intervalMs);
    signal?.addEventListener("abort", handleAbort, { once: true });
  });
}

File: lime-rs/crates/app-server-protocol/src/protocol/v0/conversation_import.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::{AgentAttachment, AgentSession};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ConversationImportSourceClient {
    #[default]
    Codex,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ConversationImportSourceStatus {
    Ready,
    Missing,
}

impl Default for ConversationImportSourceStatus {
    fn default() -> Self {
        Self::Missing
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ConversationImportThreadStatus {
    #[default]
    NotImported,
    Importing,
    Imported,
    Conflict,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ConversationImportJobStatus {
    #[default]
    Queued,
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum ConversationImportJobPhase {
    #[default]
    Queued,
    ReadingSource,
    BuildingHistory,
    PersistingHistory,
    Finalizing,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportSourceScanParams {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_client: Option<ConversationImportSourceClient>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_root: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub query: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub include_archived: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportThreadPreviewParams {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_client: Option<ConversationImportSourceClient>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_root: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_thread_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<usize>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportThreadCommitParams {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_client: Option<ConversationImportSourceClient>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_root: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_thread_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub app_id: Option<String>,
    #[serde(default)]
    pub confirmed: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub replace_existing: Option<bool>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportJobReadParams {
    pub job_id: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportSourceSummary {
    pub source_client: ConversationImportSourceClient,
    pub status: ConversationImportSourceStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_root: Option<String>,
    pub readable: bool,
    #[serde(default)]
    pub thread_count: usize,
    #[serde(default)]
    pub source_home_exists: bool,
    #[serde(default)]
    pub state_db_readable: bool,
    #[serde(default)]
    pub rollout_file_count: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub indexed_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub state_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ImportedThreadSummary {
    pub source_client: ConversationImportSourceClient,
    pub source_thread_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub created_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_provider: Option<String>,
    #[serde(default)]
    pub archived: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub import_job_id: Option<String>,
    pub import_status: ConversationImportThreadStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Value>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportSourceScanResponse {
    pub source: ConversationImportSourceSummary,
    #[serde(default)]
    pub threads: Vec<ImportedThreadSummary>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_cursor: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportSourceProvenance {
    pub source_client: ConversationImportSourceClient,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_thread_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_event_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_event_seq: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_payload_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_call_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_role: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_channel: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportFidelitySummary {
    pub messages: usize,
    pub reasoning: usize,
    pub tools: usize,
    pub commands: usize,
    pub patches: usize,
    pub approvals: usize,
    pub mcp: usize,
    pub web_search: usize,
    pub attachments: usize,
    pub unsupported: usize,
    pub provenance_only: usize,
    pub budget_dropped: usize,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportPreviewMessage {
    pub role: String,
    pub text: String,
    #[serde(default)]
    pub attachments: Vec<AgentAttachment>,
    #[serde(default)]
    pub truncated: bool,
    #[serde(default)]
    pub omitted_bytes: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provenance: Option<ConversationImportSourceProvenance>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportPreviewEvent {
    pub kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provenance: Option<ConversationImportSourceProvenance>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportPreviewDryRun {
    pub will_create_session: bool,
    pub will_append_to_existing_session: bool,
    pub will_import_messages: usize,
    pub will_import_turns: usize,
    pub will_import_timeline_items: usize,
    pub will_import_attachments: usize,
    pub unsupported_items: usize,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportPreviewSummary {
    pub line_count: usize,
    pub message_count: usize,
    pub rollout_event_items: usize,
    pub unsupported_count: usize,
    pub dry_run: ConversationImportPreviewDryRun,
    #[serde(default)]
    pub fidelity: ConversationImportFidelitySummary,
    pub truncated: bool,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportThreadPreviewResponse {
    pub source: ConversationImportSourceSummary,
    pub thread: ImportedThreadSummary,
    pub summary: ConversationImportPreviewSummary,
    #[serde(default)]
    pub messages: Vec<ConversationImportPreviewMessage>,
    #[serde(default)]
    pub events: Vec<ConversationImportPreviewEvent>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportThreadCommitResponse {
    pub session: AgentSession,
    pub thread: ImportedThreadSummary,
    pub summary: ConversationImportPreviewSummary,
    pub imported_messages: usize,
    pub imported_turns: usize,
    pub can_continue: bool,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportJobProgress {
    pub phase: ConversationImportJobPhase,
    #[serde(default)]
    pub completed_items: usize,
    #[serde(default)]
    pub total_items: usize,
    #[serde(default)]
    pub completed_turns: usize,
    #[serde(default)]
    pub total_turns: usize,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportJob {
    pub job_id: String,
    pub source_client: ConversationImportSourceClient,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_thread_id: Option<String>,
    pub status: ConversationImportJobStatus,
    pub progress: ConversationImportJobProgress,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<ConversationImportThreadCommitResponse>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportThreadCommitStartResponse {
    pub job: ConversationImportJob,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct ConversationImportJobReadResponse {
    pub job: ConversationImportJob,
}
